### ISSUE TITLE
Lint against the use of ? in places where Err is expected

### ISSUE THREAD

**Issue by Zaerei** [rust-lang/rust#45336]
I did, this, got really confused trying to fix the "erroneous error" that was happening, and then laughed when I realized what I'd done.

Minimized code with a bogus error message:

```
some_option.ok_or::<Box<Error>>(Err(
            "Did not expect None here"
                .to_string(),
        )?)? 
```

However, my colleague made the same exact error independently and it still took me a minute to see it, so it may be worth linting against.

We knew that a `String` could become a boxed error, and independently tried to make it compile, and both ended up wrapping it in an `Err` and then using `?` which uh... immediately returns it before ever calling `ok_or` causing the function to always fail. What we wanted was 

```
some_option.ok_or::<Box<Error>>(From::from(
            "Did not expect None here"
                .to_string(),
        ))? 
```

It may be worth linting against things like this somehow, because it's really hard to spot if it happens. Even after I'd already been bitten by it, it took me a few minutes to spot it in his code.

I'm not sure how general it can be, but at a first glance if a function returns `Result<T,E>` and takes in an `E` as an argument linting against using `?` on the argument in the `E` position(s) may work. But if not, linting against the use of `?` in `std` functions like `ok_or` whose purpose is to construct an error works too.

**Comment by oli-obk** [comment]
I think this would be something that clippy would lint against. Feel free to open an issue at https://github.com/rust-lang-nursery/rust-clippy/issues

**Comment by scottmcm** [comment]
Did you get any warnings?  `foo(Err(bar)?)` ought to know that `foo` is dead there...

**Comment by LinearZoetrope** [comment]
Nope, no warnings were generated.

**Comment by Mark-Simulacrum** [comment]
I think a clippy lint would be the right avenue to explore for this at least to start off, so going to go ahead and close as such.
